File: 04_Binary_Search/03_Koko_Eating_Bananas.py
class Solution(object):
    def minEatingSpeed(self, piles, h):
        """
        :type piles: List[int]
        :type h: int
        :rtype: int
        """
        left, right = 1, max(piles)
        ans = right

        while left <= right:
            mid = (left + right) // 2
            hours = 0
            for pile in piles:
                hours += math.ceil(float(pile)/mid)
            
            if hours <= h:
                right = mid - 1
                ans = min(ans, mid)
            else:
                left = mid + 1
        
        return ans

File: 04_Binary_Search/03_Koko_Eating_Bananas.cpp
class Solution
{
public:
    int minEatingSpeed(vector<int> &piles, int h)
    {
        int left = 1, right = *max_element(piles.begin(), piles.end());
        int ans = right;

        while (left <= right)
        {
            int mid = left + (right - left) / 2;
            int64_t hours = 0;

            for (int pile : piles)
                hours += (pile + mid - 1) / mid;

            if (hours <= h)
            {
                right = mid - 1;
                ans = min(ans, mid);
            }
            else
            {
                left = mid + 1;
            }
        }

        return ans;
    }
};